mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md

```diff
@@ -0,0 +1,53 @@
+
+### 2) Cross-platform
+Goes hand in hand with 1, it should work across macOS/Linux/Windows without any weird shell quirks
+
+### 3) Speed
+We should aim for orchestration to be reasonably fast. Usually npm packages will cache things like configuration flags, model files if not changed, things like that so after the first run the npm command will run significantly faster, and since one of the big use cases for Cloesce is to advance vibe coding beyond just localhost, I assume users will be running cloesce over and over.
+
+### 4) Clear lifecycle
+Multiple stages of execution, I'll go more into this later when hashing out how we want to design orchestration
+
+### 5) Strong failure semantics
+If we want to run Cloesce with one command such that with a single TS model file, we output wrangler files ready to be deployed. To do this without it being a bad experience, we need to have Cloesce fail fast with actionable errors, they should at least point the user to the right direction. There's multiple fail points that can happen with Cloesce (json generation, d1/wrangler generation, client generation, migrations) so we need to make sure that errors are verbose enough such that if something fails, the user should be able to diagnose and fix the issue in their TS model file.
+
+### 6) Excellent DX
+Potential verbose commands for logging, progress bars, `Cloesce doctor` maybe to verify the ts model file before running anything
+
+---
+
+With these design philosophies, we can go ahead and actually plan out how we want to handle npm orchestration.
+
+## Commands and Architecture
+
+The first thing I want to cover is commands. Ideally we run `cloesce run` and everything is generated, all the user has to do is `wrangler run dev`
+
+But what goes under the hood? First thing I want to make clear is that npm should not just be a underlying bash script. We should treat the build as a dataflow, not a bunch of shell commands. **Essentially we need to build a directed acyclic graph since most good npm packages use frameworks or have their own systems that do this.** Each step (TS to IR, IR to workers.ts, IR to SQL, Wrangler build, etc.) is a task node which consumes inputs and produces outputs. So the edges say "the output of this task node will feed into the next task", and when we rebuild we can detect what the user changed, find only the tasks downstream that need to be rerun, and rebuild the minimal set. This means that things will be way faster when we rebuild, we can run things via parallelism if we want to (we can generate workers AND d1 at the same time), and we have observability into what changed for logging purposes.
```

Comment:
This is the heart of incremental compilation, which I've been exploring. A naive solution is to create a merkle tree of the CIDL and DFS to find hash differences, but it will require a major rework of how our compiler works.